File: app/src/main/java/com/recursia/popularmovies/presentation/presenters/SearchScreenPresenter.java
package com.recursia.popularmovies.presentation.presenters;

import com.arellomobile.mvp.InjectViewState;
import com.arellomobile.mvp.MvpPresenter;
import com.recursia.popularmovies.domain.SearchScreenInteractor;
import com.recursia.popularmovies.domain.models.Movie;
import com.recursia.popularmovies.presentation.views.SearchScreenContract;
import com.recursia.popularmovies.utils.LangUtils;

import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.subjects.BehaviorSubject;
import io.reactivex.subjects.Subject;

@InjectViewState
public class SearchScreenPresenter extends MvpPresenter<SearchScreenContract> {
    private final static int TIMEOUT = 300;
    private final static TimeUnit TIME_UNIT = TimeUnit.MILLISECONDS;
    private final static int QUERY_PAGE = 1;
    private final SearchScreenInteractor searchScreenInteractor;
    private final CompositeDisposable compositeDisposable = new CompositeDisposable();
    private final Subject<String> subject = BehaviorSubject.create();

    public SearchScreenPresenter(SearchScreenInteractor searchScreenInteractor) {
        this.searchScreenInteractor = searchScreenInteractor;
    }

    @Override
    protected void onFirstViewAttach() {
        initLiveSearch();
    }

    private void initLiveSearch() {
        Disposable d = subject
                .debounce(TIMEOUT, TIME_UNIT, AndroidSchedulers.mainThread())
                .distinctUntilChanged()
                .subscribe(this::updateDisplayedList);
        compositeDisposable.add(d);
    }

    private void updateDisplayedList(String query) {
        Disposable d = searchScreenInteractor
                .getMoviesByQuery(query, QUERY_PAGE, LangUtils.getDefaultLanguage())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        (m) -> getViewState().setMovies(m),
                        (t) -> getViewState().showErrorMessage(t.getLocalizedMessage())
                );
        compositeDisposable.add(d);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        compositeDisposable.dispose();
    }

    public void onItemClicked(Movie item) {
        getViewState().openDetailScreen(item);
    }

    public void onQueryTextChanged(String query) {
        if (!query.isEmpty()) {
            subject.onNext(query);
        }
    }

}

File: app/src/main/java/com/recursia/popularmovies/presentation/presenters/SearchScreenPresenter.kt
package com.recursia.popularmovies.presentation.presenters

import com.recursia.popularmovies.Screens
import com.recursia.popularmovies.domain.SearchScreenInteractor
import com.recursia.popularmovies.domain.models.Movie
import com.recursia.popularmovies.presentation.views.contracts.SearchScreenContract
import com.recursia.popularmovies.utils.LangUtils
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject
import moxy.InjectViewState
import moxy.MvpPresenter
import ru.terrakok.cicerone.Router
import java.util.concurrent.TimeUnit

@InjectViewState
class SearchScreenPresenter(
    private val searchScreenInteractor: SearchScreenInteractor,
    private val router: Router
) : MvpPresenter<SearchScreenContract>() {
    private val compositeDisposable = CompositeDisposable()
    private val subject = BehaviorSubject.create<String>()
    private var isLoading = false
    private var currentPage = 1
    private var currentQuery: String? = null

    override fun onFirstViewAttach() {
        initLiveSearch()
    }

    private fun initLiveSearch() {
        val d = subject
                .debounce(TIMEOUT.toLong(), TIME_UNIT, AndroidSchedulers.mainThread())
                .distinctUntilChanged()
                .subscribe { updateDisplayedList(it) }
        compositeDisposable.add(d)
    }

    private fun updateDisplayedList(query: String) {
        val d = searchScreenInteractor
                .getMoviesByQuery(query, currentPage, LangUtils.defaultLanguage)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { m -> viewState.setMovies(m) },
                        { t -> viewState.showErrorMessage(t.localizedMessage) }
                )
        compositeDisposable.add(d)
    }

    override fun onDestroy() {
        super.onDestroy()
        compositeDisposable.clear()
    }

    fun onItemClicked(item: Movie) {
        router.navigateTo(Screens.DetailScreen(item.id))
    }

    fun onQueryTextChanged(query: String) {
        if (query.isNotEmpty()) {
            if (query != currentQuery) {
                currentPage = 1
            }
            currentQuery = query
            subject.onNext(query)
        }
    }

    fun onBackPressed() {
        router.exit()
    }

    fun bottomIsReached() {
        if (!isLoading) {
            val d = searchScreenInteractor
                    .getMoviesByQuery(currentQuery!!, currentPage + 1, LangUtils.defaultLanguage)
                    .observeOn(AndroidSchedulers.mainThread())
                    .doFinally { isLoading = false }
                    .doOnSuccess { currentPage += 1 }
                    .doOnSubscribe { isLoading = true }
                    .subscribe(
                            { viewState.addMovies(it) },
                            { viewState.showErrorMessage(it.localizedMessage) }
                    )
            compositeDisposable.add(d)
        }
    }

    companion object {
        private const val TIMEOUT = 300
        private val TIME_UNIT = TimeUnit.MILLISECONDS
    }
}
